src/uct/tcp/tcp_ep.c

```diff
@@ -1151,20 +1213,27 @@ uct_tcp_ep_am_sendv_get_trace_payload(uct_tcp_am_hdr_t *hdr,
 }
 
 static inline ucs_status_t
-uct_tcp_ep_am_sendv(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
-                    int short_sendv, uct_tcp_am_hdr_t *hdr,
+uct_tcp_ep_am_sendv(uct_tcp_ep_t *ep, int short_sendv, uct_tcp_am_hdr_t *hdr,
                     size_t send_limit, const void *header,
                     struct iovec *iov, size_t iov_cnt)
 {
+    uct_tcp_iface_t UCS_V_UNUSED *iface = ucs_derived_of(ep->super.super.iface,
+                                                         uct_tcp_iface_t);
     ucs_status_t status;
+    size_t sent_length;
 
-    ep->tx.length += hdr->length + sizeof(*hdr);
+    uct_tcp_ep_advance_tx(ep, hdr);
 
     ucs_assertv((ep->tx.length <= send_limit) &&
                 (iov_cnt > 0), "ep=%p", ep);
 
-    status = ucs_socket_sendv_nb(ep->fd, iov, iov_cnt,
-                                 &ep->tx.offset, NULL, NULL);
+    status = ucs_socket_sendv_nb(ep->fd, iov, iov_cnt, &sent_length);
+    if (ucs_unlikely((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS))) {
+        uct_tcp_ep_handle_disconnected(ep, status);
+        return status;
+    }
+
+    uct_tcp_ep_comp_tx(ep, sent_length);
```

Comment:
i don't see which lines it replaced? was it missing before?

Reply:
but comp_tx is doing something else:
```
    iface->outstanding -= sent_length;
    ep->tx.offset      += sent_length;
```

Reply:
yes, now we do
```
uct_tcp_ep_tx_started();

sendv();

uct_tcp_ep_tx_completed();
```

Reply:
so it was actually missing before this PR right?

Reply:
yes, `iface::outstanding` was never decreased in case of the error in PUT Zcopy, AM Zcopy and AM Short

Reply:
so, issues was found during testing of peer failure in our gtests

Reply:
ok